Shell completion for plugins

marckhouzam opened the pull request:
#### What type of PR is this?

/kind feature
/sig cli

#### What this PR does / why we need it:

This PR does three things:
1- includes plugin names during kubectl shell completion (cherry-pick of @superbrothers's commit from #76561)
2- allows a plugin to provide completions through an executable script named `kubectl_complete-<pluginName>` present on `$PATH`
3- adds support for shell completion to the `sample-cli-plugin`

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #74178

#### Special notes for your reviewer:

Tests are missing at the moment.

When doing completion for arguments or flags for a plugin, kubectl will call `kubectl_complete-<plugin>` to obtain the list of completions.  For example, for the `krew` plugin, when the user triggers completion by doing:
```
kubectl krew <TAB>
```
kubectl will look on `$PATH` for an executable file called `kubectl_complete-krew`.  This file should print the list of valid completions for the plugin to stdout.  kubectl will then present these completion choices to the user.  This approach is the one used by the Helm project for its plugins: https://helm.sh/docs/topics/plugins/#dynamic-completion

**Backwards-compatibility**:

The solution is backwards-compatible with plugins in both cases:
1. existing plugins will work with a new `kubectl` using this solution
2. plugins supporting this solution will work with an old `kubectl`

**Krew**:

From what I can see from `krew` (I'm no expert), there is no current way to deliver a separate executable `kubectl_complete-<plugin>` completion file.  Therefore, if this solution is chosen, it may be interesting to add support for it in `krew`.

*Explanation*: `krew` installs one plugin binary on $PATH using the plugin naming convention.  That means that the only way for a plugin completion solution to work out-of-the-box with `krew` would be to call the binary plugin itself to get completions (using a special sub-command or flag).  This is not a viable solution because if a plugin ignores arguments and flags, calling it to obtain completions will mistakenly run the actual plugin logic.

**How to test**

_Manually creating a completion script:_

_For zsh (but very similar for bash, fish or powershell)._

We need a plugin to provide completion through a new `kubectl_complete-<pluginName>` file.  Let's create one ourselves.  Let's do it for the `krew` plugin, which I assume many in the community have installed.  Since `krew` uses Cobra, it has the `__complete` command, which makes it very easy to obtain completions, as will become clear below:
```
cat <<EOF >kubectl_complete-krew
#!/usr/bin/env sh

# Call the __complete command passing it all arguments
kubectl krew __complete "\$@"
EOF

chmod u+x kubectl_complete-krew
```
_Please put the `kubectl_complete-krew` file somewhere on your $PATH_.  Then:
```
cd $GOPATH/src/k8s.io/kubernetes
go build -o /tmp/kubectl ./cmd/kubectl
source <(/tmp/kubectl completion zsh)

/tmp/kubectl <TAB>
# Notice that plugins are now listed as choices

/tmp/kubectl krew <TAB>
# Notice that krew sub-commands are now listed
```

_Automatically generating completion scripts:_

I've written a new plugin, "[plugin-completion](https://github.com/marckhouzam/kubectl-plugin_completion)", which automatically generates the proper shell completion script for installed plugins that use Cobra.  It works with more than half the plugins registered with krew. It can help test this PR quite easily.

**NOTE**: Flag completion for plugins works with Cobra 1.3 which is now part of the master branch.

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Shell completion will now show plugin names when appropriate.  Furthermore, shell completion will work for plugins that provide such support.
```

<!--

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
```docs

```
-->

superbrothers committed:
Add completion support for kubectl plugins

This commit allows us to get shell completion for kubectl plugins.

For example, with the use of a plugin named 'kubectl-krew', completing
   kubectl k<tab>
will include 'krew' in the list of possible completions.

Because the shell completion scripts are generated by Cobra, this commit
took the approach of registering every plugin as a Cobra command.  This
makes Cobra aware of each plugin command when generating completions.

For efficiency, searching for plugins and registering them as Cobra
commands is only done when needed, which is when calling the
'kubectl completion' command in the case of bash, as it includes all
commands and subcommands directly in the bash generated script.

For the other shells, this will need to be done for the
'kubectl __complete' command, in a follow-up command.

Co-authored-by: Marc Khouzam <marc.khouzam@montreal.ca>
Signed-off-by: Kazuki Suda <kazuki.suda@gmail.com>

marckhouzam committed:
Add support for shell completion for plugins

When doing completion for arguments or flags for a plugin, kubectl will
call "kubectl_complete-<plugin>" to obtain the list of completions.
For example, for "krew" plugin, when the user triggers completion using:

$ kubectl krew <tab><tab>

kubectl will look for an executable file called "kubectl_complete-krew".
This file should print the list of valid completions for the plugin to
stdout.  Using cobra.ShellCompDirective as the last line of the output
is supported as is done by Cobra.

We also clear global flags when doing plugin completion because plugins
don't necessarily accept the global kubectl flags.  If some plugins do,
they will need to include such flags in their kubectl_complete-<plugin>
output.

Signed-off-by: Marc Khouzam <marc.khouzam@montreal.ca>

Drugoy commented:
@marckhouzam Hey Marc!
I'm a bit confused about how your suggested solution works.
Did I get it right that in order to bring shell completion for my self-written kubectl plugin I would need to manually rebuild kubectl with a patch that adds some instructions on how kubectl should provide completion for my plugin?

Also, in your example - I don't see what list of values would be suggested as completions for your plugin, could you explain how it'd work?

marckhouzam commented:
> @marckhouzam Hey Marc! I'm a bit confused about how your suggested solution works. Did I get it right that in order to bring shell completion for my self-written kubectl plugin I would need to manually rebuild kubectl with a patch that adds some instructions on how kubectl should provide completion for my plugin?

Hi @Drugoy.  Currently, if you want to try the proposed solution, then you need to rebuild kubectl using this PR and then you need to provide an executable file called `kubectl_complete-<yourplugin>` which can be found on $PATH.  `kubectl` will call this file passing it the current command-line following your plugin name and will expect the file to output the correct completions for the specified command-line.

So if you take the `krew` plugin for example, and the user types `kubectl krew u<TAB>`, `kubectl` built with this PR will try to find an executable called `kubectl_complete-krew` and will call it passing it the `u` argument.  That executable file can look something like this:
```
!#/bin/bash

# Krew has the following sub-commands
echo update
echo upgrade
echo search
echo install
echo uninstall
```
This is a simple example, but you would have to also handle flags if the user typed a `-`, but also output different completions if the user has already specified the first level of subcommands on the command-line, etc

Maybe the helm documentation can help?  I'm proposing to do the same thing for kubectl plugins: https://helm.sh/docs/topics/plugins/#providing-shell-auto-completion

Drugoy commented:
@marckhouzam, thanks for explaining in details.
But what about dynamic lists that are currently being provided by kubectl completion functions?
How to reuse them for plugins?

2 simple kubectl plugin example ideas: say, I don't like to remember how to use kubectl to do a few things, so, instead, I'd like to call:
  - `kubectl context ${context}` to switch active context (i.e. that "plugin" is an alias for `kubectl config use-context ${context}` and `${context}` should be tab-completable for the plugin)
  - `kubectl namespace ${namespace}` to switch namespace for the active context (i.e. that "plugin" is an alias for `kubectl config set-context --current --namespace=${namespace}` and `${namespace}` should be tab-completable for the plugin)

In the first example kubectl parses kube-config to list existing contexts.
In the second example kubectl queries KubeAPI (of the Kubernetes cluster from active context) to list some resources like pods/deployments/namespaces/whatever.
In both examples the lists are generated/dynamic and instead of writing code do what kubectl does to show these lists - I'd like to be able to just reference kubectl's existing completion functions.
What do you think about that?

p.s.: IMO such completion cases should be treated like specials and there shouldn't be necessity to provide any extra `kubectl_complete-*` files, but instead there should be a way to just reference kubectl's completion functions somehow.

marckhouzam commented:
> @marckhouzam, thanks for explaining in details. But what about dynamic lists that are currently being provided by kubectl completion functions? How to reuse them for plugins?
> 
> * kubectl namespace ${namespace} to switch namespace for the active context (i.e. that "plugin" is an alias for kubectl config set-context --current --namespace=${namespace} and ${namespace} should be tab-completable for the plugin)

For dynamic completions your `kubectl_complete-<plugin>` file can call `kubectl` (or any other program) to obtain the list of completions.  For example, in your example above, the `kubectl_complete-namespace` script can be (I may not have gotten the syntax perfectly but it illustrates the idea):
```
#!/bin/bash
kubectl get namespaces -o name
```
Such a script will provide all namespace names as completions for that plugin.

> p.s.: IMO such completion cases should be treated like specials and there shouldn't be necessity to provide any extra `kubectl_complete-*` files, but instead there should be a way to just reference kubectl's completion functions somehow.

Where would you reference such completions function?  You need the `kubectl_complete-*` to specify that reference 🙂

Drugoy commented:
@marckhouzam
IMO it shouldn't be scripts in PATH envvar, but rather be shell functions, just like kubectl currently works.
Currently, to 'install' kubectl completions one needs to do something like `source <(kubectl completion "$(basename "${SHELL}")"` and that registers `_kubectl` function that is used for completions.
You can inspect that function's code via `which _kubectl`.

IMO, kubectl plugins completions should follow this approach and also use functions.
As far as I understand, shell doesn't really care much if a command you call is a function call or an execution of some binary (from PATH), so all that probably needs to be changed here is the name of the binary/function: IMO, just like `_kubectl` - completions for kubectl plugins should also start with an underscore.

As for your bash scripts illustrating how to provide completions for kubectl plugins - it looks like the only requisite for completion scripts/functions is that they return a list of items, where each item should be on its own line.
If that's the case, your last example is a bit incorrect, as it returns namespaces with `namespace/` static prefix.
`kubectl get namespaces --no-headers -o custom-columns="NAME:.metadata.name"` or `kubectl get namespaces -ojsonpath='{range .items[*]}{.metadata.name}{"\n"}{end}` should be used instead.

But that was an illustration that provided only a single completion list, while kubectl has multiple completion lists: some sub-commands have arguments (like `-n`/`--namespace`) that have their own completion lists!

What'd you suggest to do here?
To have a bunch of separate scripts in PATH for that?
How would that look like if my plugin would need to have `-n {TAB}` completion that should suggest namespaces and final arg completion that should suggest pods (from the namespace specified in `-n` CLI arg or from active namespace if `-n` arg was not provided)?

marckhouzam commented:
> @marckhouzam IMO it shouldn't be scripts in PATH envvar, but rather be shell functions, just like kubectl currently works.
> IMO, kubectl plugins completions should follow this approach and also use functions.

This is an interesting point.  IIUC, with such an approach, for a plugin named `namespace`, `kubectl` would call the shell function `_kubectl-namespace` to obtain the completions?  Is that what you mean?  In such a case however, how will the plugin tell `kubectl` which completion function to call and when?  Also, I assume we need to implement these functions in each of the four shell languages supported by `kubectl` (bash, zsh, fish, powershell)?

> But that was an illustration that provided only a single completion list, while kubectl has multiple completion lists: some sub-commands have arguments (like `-n`/`--namespace`) that have their own completion lists!
>
> What'd you suggest to do here? To have a bunch of separate scripts in PATH for that? How would that look like if my plugin would need to have `-n {TAB}` completion that should suggest namespaces and final arg completion that should suggest pods (from the namespace specified in `-n` CLI arg or from active namespace if `-n` arg was not provided)?

You would have to write a single smart script or program that would check the arguments passed to it and print the proper completions based on that.  Here is an example of a helm plugin for which I wrote this completion script: https://github.com/marckhouzam/helm-fullstatus/blob/master/plugin.complete

I realize it can be quite complicated to write a proper completion script depending on the complexity of the plugin, but I don't think there is anything `kubectl` can do about that for dynamic completions: only the plugins knows how to find the completions.  For static completions however I am proposing an improvement to the current PR as described here: https://helm.sh/docs/topics/plugins/#static-auto-completion.  It should make static completions very easy to implement for a plugin.

Another way to make plugin completion very easy is to use a framework that provides support for completions.  For example, plugins written in Go using the Cobra project, can provide completions much more easily; I give such an example for `krew` (which uses Cobra) in the description of this PR.  Does this make sense?

Drugoy commented:
> This is an interesting point. IIUC, with such an approach, for a plugin named `namespace`, `kubectl` would call the shell function `_kubectl-namespace` to obtain the completions? Is that what you mean?

Something like that, yeah.
> In such a case however, how will the plugin tell `kubectl` which completion function to call and when?

I don't really have an explanation: I tried to understand how `kubectl`'s shell completion works, but since shell completion plugins in general are already some kind of arcane magic to me and since `kubectl` seems to use a very clever (but not very maintainable, as it looks to me) approach of introducing more and more completion functions upon execution of the initial completion function - I simply can't answer your question.
Although I get your desire to make things simple - I think it'd be way better if you figured out how `kubectl`'s own completions work and maybe then you could extend them in an elegant way.

I see that if I call `_kubectl` directly from interactive shell it says
`_arguments:comparguments:325: can only be called from completion function`
so it seems to use `_arguments` + `_describe` functions to figure out the context of completion.

> Also, I assume we need to implement these functions in each of the four shell languages supported by `kubectl` (bash, zsh, fish, powershell)?

Well, yeah, since that's what the currently documented approach of configuring shell completions for `kubectl` <a href="https://kubernetes.io/docs/tasks/tools/install-kubectl-linux/#optional-kubectl-configurations-and-plugins">already is</a>:
> Enable shell autocompletion 
kubectl provides autocompletion support for Bash and Zsh, which can save you a lot of typing.
>
> Below are the procedures to set up autocompletion for <a href="https://kubernetes.io/docs/tasks/tools/included/optional-kubectl-configs-bash-linux/">Bash</a> and <a href="https://kubernetes.io/docs/tasks/tools/included/optional-kubectl-configs-zsh/">Zsh</a>.
...

In <a href="https://github.com/kubernetes/kubernetes/blob/dec8528abaa58a4c7fc190df2cad8848799f9891/staging/src/k8s.io/kubectl/pkg/cmd/completion/completion.go#L116-L119">the source code of kubectl</a> I actually see instructions for fish and powershell as well.

> You would have to write a single smart script or program that would check the arguments passed to it and print the proper completions based on that.
...
Does this make sense?

Maybe you are right. I see that `kubectl`'s completions also use Cobra package, but I thought there should be an easy way to re-use them without writing much of own code.

marckhouzam commented:
> I don't really have an explanation: I tried to understand how `kubectl`'s shell completion works, but since shell completion plugins in general are already some kind of arcane magic to me 

I know how you feel, it took me a very long time to understand it and I'm still learning 😉 

> and since `kubectl` seems to use a very clever (but not very maintainable, as it looks to me) approach of introducing more and more completion functions upon execution of the initial completion function

This has changed with kubectl 1.22 which introduces (through the Cobra project), the use of the `__complete` command to obtain completions.  You can try the following for example, which will complete all namespaces starting with the letter `k`, but you can complete any command-line after the `__complete` command:
```
$ kubectl __complete -n k
kube-node-lease
kube-public
kube-system
:4
Completion ended with directive: ShellCompDirectiveNoFileComp
```

> > Also, I assume we need to implement these functions in each of the four shell languages supported by `kubectl` (bash, zsh, fish, powershell)?
> 
> Well, yeah, since that's what the currently documented approach of configuring shell completions for `kubectl` [already is](https://kubernetes.io/docs/tasks/tools/install-kubectl-linux/#optional-kubectl-configurations-and-plugins)

`kubectl` supports different shells, but thanks to Cobra it only needs to implement the completion logic in Go, and Cobra takes care of handling the different shells.  Please see details in https://github.com/spf13/cobra/blob/master/shell_completions.md#dynamic-completion-of-nouns

> > You would have to write a single smart script or program that would check the arguments passed to it and print the proper completions based on that.
> > ...
> > Does this make sense?
> 
> Maybe you are right. I see that `kubectl`'s completions also use Cobra package, but I thought there should be an easy way to re-use them without writing much of own code.

You are right.  The way to do that is to call `kubectl` directly and obtain the information you need.  A shell function would not be able to do any better anyway as it would also need to call `kubectl` to get the information.

Drugoy commented:
@marckhouzam How to use `__complete` command to get a list that gets suggested to me when I hit `{Tab}` after typing `kubectl get `?
It suggests a list of existing types of resources in the cluster.
I need to know that so that I could write a simple bash script for exporting/backuping data from specific namespace.

marckhouzam commented:
> @marckhouzam How to use `__complete` command to get a list that gets suggested to me when I hit `{Tab}` after typing `kubectl get `? It suggests a list of existing types of resources in the cluster. I need to know that so that I could write a simple bash script for exporting/backuping data from specific namespace.

@Drugoy You can use `kubectl api-resources -o name` to get a list of resources in the cluster.

Drugoy commented:
Thank you!

maelvls commented:
Cobra has released the [v1.3.0](https://github.com/spf13/cobra/releases/tag/v1.3.0) which includes https://github.com/spf13/cobra/pull/1161. Hopefully kubernetes/kubernetes will migrate from v1.2.1 to v1.3.0 soon!

marckhouzam commented:
> Cobra has released the [v1.3.0](https://github.com/spf13/cobra/releases/tag/v1.3.0) which includes [spf13/cobra#1161](https://github.com/spf13/cobra/pull/1161). Hopefully kubernetes/kubernetes will migrate from v1.2.1 to v1.3.0 soon!

Moving to Cobra 1.3.0 would be very useful for this PR 👍

marckhouzam commented:
I would like to point out again that for any plugins developed using Cobra, getting shell completion will be incredibly easy since all completions can be obtained by calling such plugins with `kubectl <plugin> __complete "@$"`.
I've tried this successfully with `krew`, `get-all`, `convert`, `stern`.  Cobra 1.3.0 is required for flag name completion however.  Using Cobra's `__complete` command gave me an alternative idea.

Currently, the solution requires packaging a second file `kubectl_complete-<plugin>` that will be called to provide completions.  I'm not sure where to put that file.  Currently, the PR requires it being in the same directory as the plugin itself.  The new idea is to merge that file with the plugin executable file.

By merging the two files plugins that want shell completion will need to accept being called as `kubectl <plugin> __complete "@$"` to obtain completions.  In such a case (and therefore for every plugin using Cobra), the plugin file would need to renamed from `kubectl-<plugin>` to `kubectl_complete-<plugin>` (or some other new name) to indicate to `kubectl` that the plugin does support shell completions and that it can be called with the `__complete` command. `kubectl` would then start by looking for a `kubectl_complete-<plugin>` as a plugin executable file and if there was none, it would fall back to `kubectl-<plugin>`.  Note that renaming the file is important because `kubectl` cannot blindly call any plugin with a `__complete` command as it may trigger the plugin to do unexpected things if it does not support such a command.

I personally prefer the current approach of keeping the files separate as I fear merging them is less flexible and more importantly would cause confusion for users as different plugins may use a different name format.  I still wanted to mention the idea, to hopefully get the discussion moving forward.

I think the most important thing is making shell completion as easy as possible for plugin developers.  I would therefore love to hear the opinion of plugin developers.

/cc @corneliusweig

marckhouzam commented:
I just realized this has been broken since my last push.  Apologies for that.  This latest version works.

marckhouzam commented:
Except for missing tests, I feel this is ready.  I've simplified the code and isolated all changes to the plugin package.

I'm also pretty confident this is a good approach as a solution for plugin developers.  I say this after having investigated how git allows its plugins to do shell completion.  Kubectl's plugin solution seems to be inspired by git's plugin solution (any executable on `$PATH` named `git-<name>` will be a plugin for the main git command).  Then, for completion git basically requires that:
1. a plugin developer provide a bash function named `_git_pluginName()` which will provide all completion choices (something similar must be done for zsh)
2. the user must load that function somehow

For kubectl, we would basically have the same solution, but more flexible IMHO:
1. a plugin developer should provide an executable script (any language) which will provide all completion choices (which will automatically work for all 4 shells supported by kubectl completion)
2. the user must put that script on $PATH

marckhouzam commented:
Now that the master branch includes Cobra 1.4 (#108643), this plugin-completion PR is fully functional.

I've also written a new plugin, "[plugin-completion](https://github.com/marckhouzam/kubectl-plugin_completion)", which automatically generates the proper shell completion script for installed plugins that use Cobra.  This little plugin can help test this PR quite easily.

MartinSimango commented:
Hi @marckhouzam, I've been working on a solution for the shell completion plugin problem recently and had a look at your kubectl-plugin_completion repository. 

I have created a similar [plugin](https://github.com/MartinSimango/kubectl-plugin_completion) that currently has support for zsh and bash shells. My proposed solution edits the _kubectl completion function that is generated by the "kubectl completion SHELL_NAME" command. It modifies the function and checks if the second argument is a plugin name and if it is, it calls the completion function for that plugin (if it has one).  The completion function for plugins can be specified within a config file so plugins written in any language can add their own completion functions. 

My plugin requires no change to the kubectl tool source code and can offer support for plugins not written using cobra. Please have a look at my [repository's](https://github.com/MartinSimango/kubectl-plugin_completion) README and let me know if my solution is a viable one.

marckhouzam commented:
Hi @MartinSimango.  Nice to see your interest in getting plugin completion to work.

> I've been working on a solution for the shell completion plugin problem recently and had a look at your kubectl-plugin_completion repository.

Please keep in mind that my plugin is only meant as a hack until enough plugins have included support for shell completion themselves.  It is not meant to be a "production" shell completion solution for `kubectl` plugins.

> I have created a similar [plugin](https://github.com/MartinSimango/kubectl-plugin_completion) that currently has support for zsh and bash shells. 

I've tried your plugin and I think it is pretty cool!  It may be a workable solution until there is native `kubectl` support for plugin completion.

> Please have a look at my [repository's](https://github.com/MartinSimango/kubectl-plugin_completion) README and let me know if my solution is a viable one.

Things I think are great:
1. it can work without having any `kubectl` support
1. handles both bash and zsh
1. automatically supports Cobra plugins
1. relatively simple to use for end-users

Things that can be problematic:
1. the native completion scripts are modified to inject support for plugins; this can easily break if those scripts change in a future release of `kubectl`.  In fact, with `kubectl` v1.24, the bash script has changed.  Having native `kubectl` support for plugin completion will allow to respect the shell completion API provided by Cobra, which will provide a future-proof solution
1. more work needed to support fish and powershell 
1. IIUC, non-cobra plugins would need to provide a completion function for each of the supported shells (especially if fish and powershell are supported, since their script syntax in not compatible with bash or zsh)
1. not all plugins using Cobra support the `completion` command (e.g., cert-manager plugin will throw an error, stern plugin will possibly hang forever)
1.  for some reason, for me at least, the script generation is quite slow (`time kubectl plugin_completion plugin-completion bash` takes 5 seconds), which is problematic if it is added to `.bashrc`

That being said, your plugin does provide a nicer experience than what is currently available.

As for this PR, I will be getting back to moving it forward to hopefully have it in kubectl v1.25.  The ball is currently in my court to document the solution but I have been quite busy this last month, but I'll make time soon.

MartinSimango commented:
Hi @marckhouzam, thank you so much for taking your time to test out my plugin and for giving me productive feedback.

To address some of your points which I all agree with:

> more work needed to support fish and powershell. 

I did plan on adding support for these shells but just need to find time and learn about the scripting languages.

> IIUC, non-cobra plugins would need to provide a completion function for each of the supported shells (especially if fish and powershell are supported, since their script syntax in not compatible with bash or zsh)

Yes this is correct. This would definitely be a tedious task for people who writes plugins to do. So definitely something problematic. 

>not all plugins using Cobra support the completion command (e.g., cert-manager plugin will throw an error, stern plugin will possibly hang forever)

Thanks for catching this. This can be addressed by blacklisting these plugins from completion but I would need to blacklist another plugins with similar issues. 

> for some reason, for me at least, the script generation is quite slow (time kubectl plugin_completion plugin-completion bash takes 5 seconds), which is problematic if it is added to .bashrc

I think i can fix this issue quite quick (I added multiple loops for displaying purpopses whereas everything could be done in one loop)


>  the native completion scripts are modified to inject support for plugins; this can easily break if those scripts change in a future release of kubectl. In fact, with kubectl v1.24, the bash script has changed. Having native kubectl support for plugin completion will allow to respect the shell completion API provided by Cobra, which will provide a future-proof solution

Definitely agree with this above point. I think for now my plugin could be a temporary solution until your PR is merged in. I really like that your PR is shell agnostic and one will not to write multiple completion functions for each shell (which is a tedious task - unless using cobra). I'd love to see the documentation of the solution and hopefully the PR can be merged in soon.

rumstead commented:
Has anyone successfully used the kubectl autocomplete with their plugin? Ie, I have a plugin that has all the same completions as kubectl. I ended up doing something similar to @MartinSimango, though a little different. You missout on some file and directory completions. 

```shell
functions[_kubectl-og]=$functions[_kubectl]

_kubectl() {
    words[$words[(i)safe]]=()
    _kubectl-og "$words[*]"
}
```

MartinSimango commented:
Hi @rumstead , I will shortly fix the file and directory completions (hopefully on the weekend). I've made a change to my plugin to allow for completions for plugins that have the same completions as kubectl i.e for your plugin and for a plugin like the kubectl-allctx plugin (this change hasn't been released yet but will be released in my next release).

marckhouzam committed:
Add shell completion support to plugin example

The use of `kubectl_complete-ns` enables shell completion when calling
the plugin through `kubectl ns`.  The script only completes flag names
and namespace arguments.

Signed-off-by: Marc Khouzam <marc.khouzam@gmail.com>

marckhouzam commented:
Alright, I've finally rebased this PR and posted a PR to update the plugin KEP to describe the proposed shell-completion solution for plugins: https://github.com/kubernetes/enhancements/pull/3496

I've also added to this PR support for shell completion for the `sample-cli-plugin` which could serve as an example.

Hopefully we can get this in to 1.26 🤞

marckhouzam commented:
/remove-lifecycle stale

marckhouzam commented:
/retest

soltysh reviewed (approved):
Some minor nits here and there, but all looks great - thank you! And apologies for the long delays :sweat_smile: 

/lgtm
/approve

soltysh commented:
/triage accepted
/priority backlog

superbrothers commented:
@marckhouzam Great work! 👏 🎉
